File: programcreek/python/01-09-4sum.py



def fourSum(nums, target) :

    nums.sort()

    hashSet = dict()
    result = list()

    for i in range(0, len(nums), 1):
        for j in range(i + 1, len(nums), 1):
            k = j + 1
            l = len(nums) - 1

            # print "%d %d %d %d" % (i, j, k, l)
            while k < l:
                total = nums[i] + nums[j] + nums[k] + nums[l]


                if total > target:
                    l -= 1
                elif total < target:
                    k += 1
                elif total == target:
                    temp = [nums[i], nums[j], nums[k], nums[l]]
                    tempKey = '|'.join([ str(x) for x in temp ])

                    if tempKey not in hashSet:
                        hashSet[tempKey] = 1
                        result.append(temp)
                    k += 1
                    l -= 1



    return result



print fourSum([1, 0, -1, 0, -2, 2], 0)

File: programcreek/js/01-09-4sum.js

function fourSum(nums, target) {
    var result = []
    var done = {}

    nums.sort();

    for ( var i = 0; i < nums.length; i ++ ) {
        for ( var j = i + 1; j < nums.length; j ++ ) {
            var k = j + 1;
            var l = nums.length -1;

            while (k < l) {
                var total = nums[i] + nums[j] + nums[k] + nums[l];
                if (total > target) {
                    l --;
                }
                else if (total < target) {
                    k ++;
                }
                else {

                    var temp = [nums[i], nums[j], nums[k], nums[l]];
                    var tempKey = temp.join('|');

                    if (! (tempKey in done)) {
                        done[tempKey] = 1
                        result.push(temp)
                        k ++;
                        l --;
                    }
                }
            }
        }
    }

    return result;
}




console.log( fourSum([1, 0, -1, 0, -2, 2], 0) );